### PR TITLE
Space - Lee

### PR DESCRIPTION
**Pull request by theomoondev**
# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | In the Work model, I wrote a custom method called self.spotlight, which performs a SQL database query to retrieve the Work with the highest number of votes. The database query involves joining the Works table with the Votes table, grouping by `works.id` so Works associated with multiple Vote records (i.e., by different users) only appear once, and ordering the resulting list in descending order by the `votes_count` column (which is automatically incremented anytime the Work is upvoted due to counter caching) and `votes.created_at`. The method finally returns the first element in this list, which represents the Work with the highest number of votes that was most recently upvoted.
Describe how you approached testing that model method. What edge cases did you come up with? | I wanted to make sure the method would work as expected even if there were no Works or Votes in the database, and if more than one Work has the highest number of votes. If there are no Works or Votes in the database, the method returns nil. If there is a tie between highest number of votes, it returns the Work that was most recently upvoted. To test the tie-handling logic, I created test fixtures with three unique works and a varying amount of votes. I then created a new vote for each work until there was a tie, then checked whether the method returned the last upvoted Work.
What are `session` and `flash`? What is the difference between them? | Data in the `session` hash is stored in the browser's cookies, which will last as long as they are not cleared (either by setting the session value to nil, manually clearing the browser cookies, or closing the browser). Data in the `flash` hash will only last for one more request (e.g., render or redirect); after that, they are automatically cleared. Both hashes provide ways to store data that will persist beyond the current request.
What was one thing that you gained more clarity on through this assignment? | How to perform more complicate SQL database queries to increase the efficiency of business logic methods. Whether I should put methods in the models or controllers (including ApplicationController methods that can be accessed by all controllers). I'm still unsure as to whether I should create a view partial vs. moving it to a view helper method, but I did gain more clarity on the advantage of being able to test view outputs.
What is the Heroku URL of your deployed application? | 1) master branch: https://lgbtq-media-ranker.herokuapp.com/ 2) clone-original-styling branch: https://radiant-reef-71518.herokuapp.com/

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+          <li class="nav-item app-header__nav_item">
+            <%= link_to "View All Media", works_path, class: "nav-link" %>
+          </li>
+          <li class="nav-item app-header__nav_item">
+            <%= link_to "Add a Work", new_work_path, class: "nav-link" %>
+          </li>
+          <li class="nav-item app-header__nav_item">
+            <%= link_to "View All Users", users_path, class: "nav-link" %>
+          </li>
+        </ul>
```
Comment (dHelmgren) [inline app/views/layouts/application.html.erb:33]: Nice work!

**Comment by dHelmgren** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) |
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | -
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention |
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |
